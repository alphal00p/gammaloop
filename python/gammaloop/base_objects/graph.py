from __future__ import annotations

import gammaloop.base_objects.model as model
from gammaloop.misc.common import *

# Abstract class for all vertex info
class VertexInfo(object):
    pass

class InteractonVertexInfo(VertexInfo):
    def __init__(self, vertex_rule: model.VertexRule):
        self.vertex_rule: model.VertexRule = vertex_rule

class Vertex(object):
    def __init__(self, name: str, vertex_info: VertexInfo, edges: list[Edge] | None):
        self.name: name = name
        self.vertex_rule: VertexInfo = vertex_info
        self.edges: list[Edge] = edges

class Edge(object):
    def __init__(self, name: str, particle: model.Particle, vertices: (Vertex, Vertex) | None):
        self.id: name = name
        self.particle: model.Particle = particle
        self.vertices: (Vertex, Vertex) = vertices

# TODO create serializable counterpart and export to yaml, similarly as done for the model.

class Graph(object):
    def __init__(self, name: str, vertices: list[Vertex], edges: list[Edge]):
        self.name: str = name
        self.vertices: list[Vertex] = vertices
        self.edges: list[Edge] = edges
        self.name_to_position: dict[str,dict[str,int]] = {}
    
    def get_edge(self, edge_name: str) -> Edge:
        return self.edges[self.name_to_position['edges'][edge_name]]
    
    def get_vertex(self, vertex_name: str) -> Parameter:
        return self.vertices[self.name_to_position['vertices'][vertex_name]]

    def synchronize_name_map(self) -> None:
        for i_e, edge in enumerate(self.edges):
            self.name_to_position['edges'][edge.name] = i_e
        for i_v, vertex in enumerate(self.vertices):
            self.name_to_position['vertices'][vertex.name] = i_v


    @staticmethod
    def from_qgraph(model: model.Model, qgraph_object) -> Graph:
        """ Imports graph form a stylicized qgraph python output file. Will be deprecated when using in-house gammaloop graph generation. """
        logger.info("Importing graph from qgraph python output")
        # TODO
        return Graph('default',[],[])