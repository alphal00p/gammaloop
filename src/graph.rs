use crate::model;
use ahash::RandomState;
use color_eyre::{Help, Report};
use enum_dispatch::enum_dispatch;
use eyre::eyre;
use serde::{Deserialize, Serialize};
use smartstring::{LazyCompact, SmartString};
use std::{collections::HashMap, sync::Arc};

#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
pub enum EdgeType {
    #[serde(rename = "in")]
    Incoming,
    #[serde(rename = "out")]
    Outgoing,
    #[default]
    #[serde(rename = "virtual")]
    Virtual,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[allow(non_snake_case)]
#[serde(tag = "type")]
pub enum SerializableVertexInfo {
    #[serde(rename = "external_vertex_info")]
    ExternalVertexInfo(SerializableExternalVertexInfo),
    #[serde(rename = "interacton_vertex_info")]
    InteractonVertexInfo(SerializableInteractionVertexInfo),
}

#[enum_dispatch]
pub trait HasVertexInfo {
    fn get_type(&self) -> SmartString<LazyCompact>;
}

#[derive(Debug, Clone)]
#[enum_dispatch(HasVertexInfo)]
pub enum VertexInfo {
    ExternalVertexInfo(ExternalVertexInfo),
    InteractonVertexInfo(InteractionVertexInfo),
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct SerializableInteractionVertexInfo {
    vertex_rule: SmartString<LazyCompact>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct SerializableExternalVertexInfo {
    direction: EdgeType,
    particle: SmartString<LazyCompact>,
}

#[derive(Debug, Clone)]
pub struct ExternalVertexInfo {
    direction: EdgeType,
    particle: Arc<model::Particle>,
}

impl HasVertexInfo for ExternalVertexInfo {
    fn get_type(&self) -> SmartString<LazyCompact> {
        SmartString::<LazyCompact>::from("external_vertex_info")
    }
}

#[derive(Debug, Clone)]
pub struct InteractionVertexInfo {
    #[allow(unused)]
    pub vertex_rule: Arc<model::VertexRule>,
}

impl HasVertexInfo for InteractionVertexInfo {
    fn get_type(&self) -> SmartString<LazyCompact> {
        SmartString::<LazyCompact>::from("interacton_vertex_info")
    }
}

fn serializable_vertex_info(vertex_info: &VertexInfo) -> SerializableVertexInfo {
    match vertex_info {
        VertexInfo::ExternalVertexInfo(external_vertex_info) => {
            SerializableVertexInfo::ExternalVertexInfo(SerializableExternalVertexInfo {
                direction: external_vertex_info.direction.clone(),
                particle: external_vertex_info.particle.name.clone(),
            })
        }
        VertexInfo::InteractonVertexInfo(interaction_vertex_info) => {
            SerializableVertexInfo::InteractonVertexInfo(SerializableInteractionVertexInfo {
                vertex_rule: interaction_vertex_info.vertex_rule.name.clone(),
            })
        }
    }
}

fn deserialize_vertex_info(
    model: &model::Model,
    serialized_vertex_info: &SerializableVertexInfo,
) -> VertexInfo {
    match serialized_vertex_info {
        SerializableVertexInfo::ExternalVertexInfo(external_vertex_info) => {
            VertexInfo::ExternalVertexInfo(
                ExternalVertexInfo {
                    direction: external_vertex_info.direction.clone(),
                    particle: model.get_particle(&external_vertex_info.particle),
                }
                .clone(),
            )
        }
        SerializableVertexInfo::InteractonVertexInfo(interaction_vertex_info) => {
            VertexInfo::InteractonVertexInfo(InteractionVertexInfo {
                vertex_rule: model
                    .get_vertex_rule(&interaction_vertex_info.vertex_rule)
                    .clone(),
            })
        }
    }
}

#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq)]
pub struct SerializableEdge {
    name: SmartString<LazyCompact>,
    edge_type: EdgeType,
    particle: SmartString<LazyCompact>,
    vertices: [SmartString<LazyCompact>; 2],
}

impl SerializableEdge {
    pub fn from_edge(graph: &Graph, edge: &Edge) -> SerializableEdge {
        SerializableEdge {
            name: edge.name.clone(),
            edge_type: edge.edge_type.clone(),
            particle: edge.particle.name.clone(),
            vertices: [
                graph.vertices[edge.vertices[0]].name.clone(),
                graph.vertices[edge.vertices[1]].name.clone(),
            ],
        }
    }
}

#[derive(Debug, Clone)]
pub struct Edge {
    pub name: SmartString<LazyCompact>,
    pub edge_type: EdgeType,
    pub particle: Arc<model::Particle>,
    pub vertices: [usize; 2],
}

impl Edge {
    pub fn from_serializable_edge(
        model: &model::Model,
        graph: &Graph,
        serializable_edge: &SerializableEdge,
    ) -> Edge {
        Edge {
            name: serializable_edge.name.clone(),
            edge_type: serializable_edge.edge_type.clone(),
            particle: model.get_particle(&serializable_edge.particle),
            vertices: [
                graph
                    .get_vertex_position(&serializable_edge.vertices[0])
                    .unwrap(),
                graph
                    .get_vertex_position(&serializable_edge.vertices[1])
                    .unwrap(),
            ],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SerializableVertex {
    name: SmartString<LazyCompact>,
    vertex_info: SerializableVertexInfo,
    edges: Vec<SmartString<LazyCompact>>,
}

impl SerializableVertex {
    pub fn from_vertex(graph: &Graph, vertex: &Vertex) -> SerializableVertex {
        SerializableVertex {
            name: vertex.name.clone(),
            vertex_info: serializable_vertex_info(&vertex.vertex_info),
            edges: vertex
                .edges
                .iter()
                .map(|e| graph.edges[*e].name.clone())
                .collect(),
        }
    }
}

#[derive(Debug, Clone)]
pub struct Vertex {
    pub name: SmartString<LazyCompact>,
    pub vertex_info: VertexInfo,
    pub edges: Vec<usize>,
}

impl Vertex {
    pub fn from_serializable_vertex(model: &model::Model, vertex: &SerializableVertex) -> Vertex {
        Vertex {
            name: vertex.name.clone(),
            vertex_info: deserialize_vertex_info(model, &vertex.vertex_info),
            // This will be filled in later during deserialization of the complete graph
            edges: vec![],
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct SerializableGraph {
    name: SmartString<LazyCompact>,
    vertices: Vec<SerializableVertex>,
    edges: Vec<SerializableEdge>,
    overall_factor: f64,
    external_connections: Vec<(
        Option<SmartString<LazyCompact>>,
        Option<SmartString<LazyCompact>>,
    )>,
    loop_momentum_basis: Vec<SmartString<LazyCompact>>,
    edge_signatures: Vec<(Vec<isize>, Vec<isize>)>,
}

impl SerializableGraph {
    pub fn from_graph(graph: &Graph) -> SerializableGraph {
        SerializableGraph {
            name: graph.name.clone(),
            vertices: graph
                .vertices
                .iter()
                .map(|v| SerializableVertex::from_vertex(graph, v))
                .collect(),
            edges: graph
                .edges
                .iter()
                .map(|e| SerializableEdge::from_edge(graph, e))
                .collect(),
            overall_factor: graph.overall_factor,
            external_connections: graph
                .external_connections
                .iter()
                .map(|(v1, v2)| {
                    (
                        v1.as_ref().map(|&v| graph.vertices[v].name.clone()),
                        v2.as_ref().map(|&v| graph.vertices[v].name.clone()),
                    )
                })
                .collect(),
            loop_momentum_basis: graph
                .loop_momentum_basis
                .iter()
                .map(|&e| graph.edges[e].name.clone())
                .collect(),
            edge_signatures: graph.edge_signatures.clone(),
        }
    }

    pub fn from_yaml_str(yaml_str: String) -> Result<SerializableGraph, Report> {
        serde_yaml::from_str(yaml_str.as_str())
            .map_err(|e| eyre!(format!("Error parsing graph yaml: {}", e)))
            .suggestion("Is it a correct yaml file")
    }
}

#[derive(Debug, Clone)]
pub struct Graph {
    pub name: SmartString<LazyCompact>,
    pub vertices: Vec<Vertex>,
    pub edges: Vec<Edge>,
    pub edge_signatures: Vec<(Vec<isize>, Vec<isize>)>,
    pub overall_factor: f64,
    pub external_connections: Vec<(Option<usize>, Option<usize>)>,
    pub loop_momentum_basis: Vec<usize>,
    pub vertex_name_to_position: HashMap<SmartString<LazyCompact>, usize, RandomState>,
    pub edge_name_to_position: HashMap<SmartString<LazyCompact>, usize, RandomState>,
}

impl Graph {
    pub fn from_serializable_graph(model: &model::Model, graph: &SerializableGraph) -> Graph {
        // First build vertices
        let mut vertices: Vec<Vertex> = vec![];
        let mut vertex_name_to_position: HashMap<SmartString<LazyCompact>, usize, RandomState> =
            HashMap::default();
        for vertex in &graph.vertices {
            let vertex = Vertex::from_serializable_vertex(model, vertex);
            vertex_name_to_position.insert(vertex.name.clone(), vertices.len());
            vertices.push(vertex);
        }

        let mut g = Graph {
            name: graph.name.clone(),
            vertices,
            edges: vec![],
            overall_factor: graph.overall_factor,
            external_connections: vec![],
            loop_momentum_basis: vec![],
            vertex_name_to_position,
            edge_name_to_position: HashMap::default(),
            edge_signatures: graph.edge_signatures.clone(),
        };

        // Then build edges
        g.edges = graph
            .edges
            .iter()
            .map(|e| Edge::from_serializable_edge(model, &g, e))
            .collect();
        for (i_e, e) in g.edges.iter().enumerate() {
            g.edge_name_to_position.insert(e.name.clone(), i_e);
        }

        g.external_connections = graph
            .external_connections
            .iter()
            .map(|(v1, v2)| {
                (
                    v1.as_ref().map(|v| {
                        g.get_vertex_position(v).unwrap_or_else(|| {
                            panic!("Cannot find vertex named {} in this graph.", v)
                        })
                    }),
                    v2.as_ref().map(|v| {
                        g.get_vertex_position(v).unwrap_or_else(|| {
                            panic!("Cannot find vertex named {} in this graph.", v)
                        })
                    }),
                )
            })
            .collect();

        g.loop_momentum_basis = graph
            .loop_momentum_basis
            .iter()
            .map(|e| g.get_edge_position(e).unwrap())
            .collect();

        g
    }

    #[inline]
    pub fn get_vertex(&self, name: &SmartString<LazyCompact>) -> Option<&Vertex> {
        match self.vertex_name_to_position.get(name) {
            Some(position) => Some(&self.vertices[*position]),
            None => None,
        }
    }

    #[inline]
    pub fn get_vertex_position(&self, name: &SmartString<LazyCompact>) -> Option<usize> {
        self.vertex_name_to_position.get(name).map(|p| *p)
    }

    #[inline]
    pub fn get_edge(&self, name: &SmartString<LazyCompact>) -> Option<&Edge> {
        match self.edge_name_to_position.get(name) {
            Some(position) => Some(&self.edges[*position]),
            None => None,
        }
    }

    #[inline]
    pub fn get_edge_position(&self, name: &SmartString<LazyCompact>) -> Option<usize> {
        self.edge_name_to_position.get(name).map(|p| *p)
    }
}
