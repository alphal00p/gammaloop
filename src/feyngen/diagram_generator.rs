use log::debug;

use super::{FeynGenError, FeynGenOptions};
use crate::{graph::BareGraph, model::Model};
use symbolica::graph::Graph as SymbolicaGraph;

pub struct FeynGen {
    pub options: FeynGenOptions,
}

impl FeynGen {
    pub fn new(options: FeynGenOptions) -> Self {
        Self { options }
    }

    pub fn generate(&self, model: &Model) -> Result<Vec<BareGraph>, FeynGenError> {
        debug!(
            "Generating Feynman diagrams for model {} and process:\n{}",
            model.name, self.options
        );

        let vertex_signatures = model
            .vertex_rules
            .iter()
            .map(|v| {
                v.particles
                    .iter()
                    .map(|p| p.name.as_str())
                    .collect::<Vec<_>>()
            })
            .collect::<Vec<_>>();

        let mut external_edges = self
            .options
            .initial_pdgs
            .iter()
            .enumerate()
            .map(|(i_initial, pdg)| {
                (
                    i_initial + 1,
                    model
                        .get_particle_from_pdg(*pdg as isize)
                        .clone()
                        .name
                        .clone(),
                )
            })
            .collect::<Vec<_>>();
        external_edges.extend(
            self.options
                .final_pdgs
                .iter()
                .enumerate()
                .map(|(i_final, pdg)| {
                    (
                        self.options.initial_pdgs.len() + i_final + 1,
                        model
                            .get_particle_from_pdg(*pdg as isize)
                            .clone()
                            .name
                            .clone(),
                    )
                })
                .collect::<Vec<_>>(),
        );
        debug!("external_edges = {:?}", external_edges);
        debug!("vertex_signatures = {:?}", vertex_signatures);

        let graphs = SymbolicaGraph::generate(
            &external_edges
                .iter()
                .map(|(i, name)| (*i, name.as_str()))
                .collect::<Vec<_>>(),
            &vertex_signatures,
            None,
            Some(self.options.loop_count_range.1),
            self.options.filters.get_max_bridge(),
            self.options.filters.allow_tadpoles(),
        );

        debug!("Symbolica generated {} graphs", graphs.len());

        Ok(vec![])
    }
}
