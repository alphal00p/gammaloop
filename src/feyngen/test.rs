use std::sync::Arc;

use ahash::HashMap;
use ahash::HashMapExt;
use symbolica::graph::Graph;

use crate::feyngen::diagram_generator::EdgeColor;
use crate::model::ColorStructure;
use crate::model::VertexRule;
use crate::tests_from_pytest::load_generic_model;

use super::diagram_generator::NodeColorWithVertexRule;
use super::GenerationType;
use super::{diagram_generator::FeynGen, FeynGenFilter, FeynGenFilters, FeynGenOptions};

#[test]
fn cut_content() {
    let model = load_generic_model("sm");

    let mut coupling = HashMap::new();
    coupling.insert("QED".into(), 6);
    let mut pert = HashMap::new();
    pert.insert("QCD".into(), 1);
    let filters = FeynGen::new(FeynGenOptions {
        generation_type: GenerationType::CrossSection,
        initial_pdgs: vec![-11, 11],
        final_pdgs: vec![5, -5, 25],
        loop_count_range: (3, 3),
        symmetrize_initial_states: true,
        symmetrize_final_states: true,
        symmetrize_left_right_states: true,
        allow_symmetrization_of_external_fermions_in_amplitudes: false,
        max_multiplicity_for_fast_cut_filter: 6,
        amplitude_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
        ]),
        cross_section_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
            FeynGenFilter::PerturbativeOrders(pert),
            FeynGenFilter::CouplingOrders(coupling),
            FeynGenFilter::LoopCountRange((3, 3)),
        ]),
    });

    let mut graph = Graph::new();
    #[allow(non_snake_case)]
    let bbH = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_78".into()),
    };
    let bba = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_73".into()),
    };
    let bbg = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_76".into()),
    };
    let epema = NodeColorWithVertexRule {
        external_tag: 0,
        vertex_rule: model.get_vertex_rule(&"V_98".into()),
    };

    let dummy_external_vertex_rule = Arc::new(VertexRule {
        name: "external".into(),
        couplings: vec![],
        lorentz_structures: vec![],
        particles: vec![],
        color_structures: ColorStructure::new(vec![]),
    });

    let v0 = graph.add_node(bbH.clone());
    let v1 = graph.add_node(bbH.clone());
    let v2 = graph.add_node(bbg.clone());
    let v3 = graph.add_node(bba.clone());
    let v4 = graph.add_node(bbg.clone());
    let v5 = graph.add_node(bba.clone());
    let v6 = graph.add_node(epema.clone());
    let v7 = graph.add_node(epema.clone());
    let e1 = NodeColorWithVertexRule {
        external_tag: 1,
        vertex_rule: dummy_external_vertex_rule.clone(),
    };
    let mut e2 = e1.clone();
    e2.external_tag = 2;
    let mut e3 = e1.clone();
    e3.external_tag = 3;
    let mut e4 = e1.clone();
    e4.external_tag = 4;

    let v8 = graph.add_node(e1.clone());
    let v9 = graph.add_node(e2.clone());
    let v10 = graph.add_node(e3.clone());
    let v11 = graph.add_node(e4.clone());

    let h = EdgeColor::from_particle(model.get_particle(&"H".to_string().into()));

    let b = EdgeColor::from_particle(model.get_particle(&"b".to_string().into()));
    let bbar = EdgeColor::from_particle(
        model
            .get_particle(&"b".to_string().into())
            .get_anti_particle(&model),
    );
    let g = EdgeColor::from_particle(model.get_particle(&"g".to_string().into()));
    let a = EdgeColor::from_particle(model.get_particle(&"a".to_string().into()));
    let eplus = EdgeColor::from_particle(model.get_particle(&"e+".to_string().into()));
    let eminus = EdgeColor::from_particle(model.get_particle(&"e-".to_string().into()));
    graph.add_edge(v0, v1, true, h).unwrap();
    graph.add_edge(v0, v5, true, b).unwrap();
    graph.add_edge(v1, v3, true, b).unwrap();
    graph.add_edge(v2, v4, true, g).unwrap();
    graph.add_edge(v2, v4, true, b).unwrap();
    graph.add_edge(v3, v2, true, b).unwrap();
    graph.add_edge(v3, v7, true, a).unwrap();
    graph.add_edge(v4, v1, true, b).unwrap();
    graph.add_edge(v5, v0, true, b).unwrap();
    graph.add_edge(v5, v6, true, a).unwrap();
    graph.add_edge(v6, v10, false, eplus).unwrap();
    graph.add_edge(v6, v11, false, eminus).unwrap();
    graph.add_edge(v8, v7, false, eplus).unwrap();
    graph.add_edge(v9, v7, false, eminus).unwrap();

    let (n_unresolved, unresolved_type) = filters.unresolved_cut_content(&model);
    assert!(!filters.contains_cut(&model, &graph, n_unresolved, &unresolved_type));

    let mut double_double_triangle = Graph::new();
    let v0 = double_double_triangle.add_node(e1.clone());
    let v1 = double_double_triangle.add_node(e2.clone());
    let v2 = double_double_triangle.add_node(e3.clone());
    let v3 = double_double_triangle.add_node(e4.clone());

    let v4 = double_double_triangle.add_node(bba.clone());
    let v5 = double_double_triangle.add_node(bba.clone());
    let v6 = double_double_triangle.add_node(bbH.clone());
    let v7 = double_double_triangle.add_node(bbH.clone());
    let v8 = double_double_triangle.add_node(bbg.clone());
    let v9 = double_double_triangle.add_node(bbg.clone());
    let v10 = double_double_triangle.add_node(bbg.clone());
    let v11 = double_double_triangle.add_node(bbg.clone());
    let v12 = double_double_triangle.add_node(epema.clone());
    let v13 = double_double_triangle.add_node(epema.clone());

    double_double_triangle
        .add_edge(v0, v13, true, eplus)
        .unwrap();
    double_double_triangle
        .add_edge(v1, v13, true, eminus)
        .unwrap();
    double_double_triangle.add_edge(v4, v13, false, a).unwrap();
    double_double_triangle
        .add_edge(v4, v11, true, bbar)
        .unwrap();
    double_double_triangle.add_edge(v4, v10, true, b).unwrap();
    double_double_triangle.add_edge(v10, v11, false, g).unwrap();
    double_double_triangle.add_edge(v6, v11, true, b).unwrap();
    double_double_triangle
        .add_edge(v6, v10, true, bbar)
        .unwrap();
    double_double_triangle.add_edge(v6, v7, true, h).unwrap();
    double_double_triangle.add_edge(v7, v8, true, bbar).unwrap();
    double_double_triangle.add_edge(v7, v9, true, b).unwrap();
    double_double_triangle.add_edge(v5, v8, true, b).unwrap();
    double_double_triangle.add_edge(v5, v9, true, bbar).unwrap();
    double_double_triangle.add_edge(v8, v9, true, g).unwrap();
    double_double_triangle.add_edge(v5, v12, true, a).unwrap();
    double_double_triangle
        .add_edge(v12, v2, true, eminus)
        .unwrap();
    double_double_triangle
        .add_edge(v12, v3, true, eplus)
        .unwrap();

    let mut coupling = HashMap::new();
    coupling.insert("QED".into(), 6);
    let mut pert = HashMap::new();
    pert.insert("QCD".into(), 2);
    let filters = FeynGen::new(FeynGenOptions {
        generation_type: GenerationType::CrossSection,
        initial_pdgs: vec![-11, 11],
        final_pdgs: vec![5, -5, 25],
        loop_count_range: (4, 4),
        symmetrize_initial_states: true,
        symmetrize_final_states: true,
        symmetrize_left_right_states: true,
        allow_symmetrization_of_external_fermions_in_amplitudes: false,
        max_multiplicity_for_fast_cut_filter: 6,
        amplitude_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
        ]),
        cross_section_filters: FeynGenFilters(vec![
            FeynGenFilter::SelfEnergyFilter(Default::default()),
            FeynGenFilter::ParticleVeto(vec![
                23, 24, 9000001, 9000002, 9000003, 9000004, 12, 14, 16, 2, 4, 6, 3, 250, 251, 13,
                15,
            ]),
            FeynGenFilter::TadpolesFilter(Default::default()),
            FeynGenFilter::ZeroSnailsFilter(Default::default()),
            FeynGenFilter::PerturbativeOrders(pert),
            FeynGenFilter::CouplingOrders(coupling),
            FeynGenFilter::LoopCountRange((4, 4)),
        ]),
    });

    let (n_unresolved, unresolved_type) = filters.unresolved_cut_content(&model);
    assert!(!filters.contains_cut(
        &model,
        &double_double_triangle,
        n_unresolved,
        &unresolved_type
    ));
}
