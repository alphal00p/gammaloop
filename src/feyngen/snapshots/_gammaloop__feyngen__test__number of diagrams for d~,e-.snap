---
source: src/feyngen/test.rs
expression: diagrams.len()
---
6
